tests/python/unittest/test_meta_schedule_tune_relay.py

```diff
@@ -323,6 +325,222 @@ def get_output(data, lib):
+
+    extracted_tasks = extract_task_from_relay(relay_mod, target, params)
+
+    tune_tasks = list(
+        filter(
+            lambda task: "dense" in task.task_name,
+            extracted_tasks,
+        )
+    )
+
+    with tempfile.TemporaryDirectory() as work_dir:
+        if do_tune:
+            config = ReplayTraceConfig(
+                num_trials_per_iter=64,
+                num_trials_total=64,
+            )
+            database = tune_extracted_tasks(
+                tune_tasks, target, config, work_dir=work_dir, postprocs=lambda: []
+            )
+        else:
+            database = JSONDatabase(
+                path_workload=osp.join(work_dir, "database_workload.json"),
+                path_tuning_record=osp.join(work_dir, "database_tuning_record.json"),
+            )
+
+            for task in tune_tasks:
+                mod = Parse._mod(task.dispatched[0])
+                workload = database.commit_workload(mod)
+
+                sch = tvm.tir.Schedule(mod)
+                block = sch.get_block("compute")
+                schedule_rule = sch.get(block).annotations["schedule_rule"]
+
+                if "dense_vnni" in schedule_rule:
+                    schedule_dense(block, M, False, sch)
+
+                tune_rec = TuningRecord(sch.trace, [0.0], workload, tvm.target.Target(target), [])
+
+                database.commit_tuning_record(tune_rec)
+
+    with ApplyHistoryBest(database):
+        with tvm.transform.PassContext(
+            opt_level=3,
+            config={"relay.backend.use_meta_schedule": True},
+        ):
+            """
+            The log should say
+            meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_expand_dims
+            meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast
+            meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast_1
+            meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_nn_batch_matmul
+
+            This means batch matmul and others are scheduled by TE, and dense (the one not warned) is found in the
+            meta schedule tuning database during ApplyHistoryBest
+            """
+            lib = relay.build(relay_mod, target=target, params=params)
+
+    runtime = tvm.contrib.graph_executor.GraphModule(lib["default"](dev))
+
+    runtime.set_input("data", data)
+    runtime.run()
+
+    out = runtime.get_output(0).numpy()
+
+    np.testing.assert_equal(out, ref)
+
+
+@pytest.mark.skip("Requires cascadelake")
+def test_tune_relay_manual_tir_vnni():
+    tir.TensorIntrin.register(VNNI_INTRIN, dot_product_desc, dot_product_intrin)
+
+    manual_tir_common(do_tune=False)
+
+    def schedule_rule_dense_vnni(sch, block):
+        schedule_dense(block, None, True, sch)
+        return [sch]
+
+    register_func("meta_schedule.dense_vnni", schedule_rule_dense_vnni)
```

Comment:
Is this overwriting the default behavior for dense on cascadelake? If so, a quick comment saying so wouldnt hurt.

Reply:
It is not "overwriting" in the sense that (1) TE scheduling is not affected and (2) without this registration, meta schedule tuning fails with 
```
  File "/home/masa/projects/dev/tvm/src/meta_schedule/space_generator/post_order_apply.cc", line 149
ValueError: Check failed: (f) is false: Custom schedule rule not found: meta_schedule.dense_vnni
```

This is because, for all TE compute annotated with `schedule_rule` like https://github.com/apache/tvm/blob/ce335c3a74185df6cc1152e53c60695d8a418d8e/python/tvm/topi/x86/dense.py#L299, we currently require the corresponding schedule rule to be registered (which this line does).

Thinking about this now, I wonder if failing is the desired behavior. Since if we don't find the custom schedule registered, we can ignore the `schedule_relu` annotation and apply automatic scheduling. cc @junrushao1994

Reply:
I want to freely add `schedule_rule` annotations to various TE compute to experiment with things, so requiring that all `schedule_rule` annotations to have the corresponding packed func registered sounds like a heavy-weight requirement to me. If we have TOPI equivalent for TIR manual schedules, such requirement is easy to satisfy, but until then I expect that manual TIR scheduling is done in an one-off fashion like this PR.

Also, it is totally reasonable to want to auto-schedule TE compute annotated with `schedule_rule`. Currently I annotated TE x86 `dense` and `batch_matmul` compute with VNNI-specific schedule rules (like "meta_schedule.dense_vnni" above) to apply my manual TIR schedule, but that prevents any automatic scheduling from happening on these TE compute. In the future when auto-tensorization is ready, I want to freely switch between manual and automatic scheduling.   

So I want "the need to annotate `schedule_relu`" and "whether or not I want to register my custom schedule rule" be decoupled.

Reply:
@masahi Thanks for the comprehensive discussion about API design and user intention!

Looks like we have 3 different proposals:
- P1. Always annotate schedule_rule even if it's not registered; Print a warning if the schedule_rule isn't registered;
- P2. Always make sure schedule_rule exists when annotating, and error out if it's not;
- P3. Always annotate schedule_rule even if it's not registered; No warning if the schedule_rule isn't registered;

Masa and I both agree that P3 may not be ideal, because silently ignoring abnormality may not be the best user experience.

> In the future when auto-tensorization is ready, I want to freely switch between manual and automatic scheduling.

Totally agree about future possibility of switching around! Definitely it's going to be a lot of fun :-)

Therefore, I would conclude that emitting a warning is probably both @masahi and I agree on. Additionally, we might enhance the search space generation to selectively check if the schedule_rule is allowed with a target-specific allowlist, but it's probably not high-priority for now.